### PR TITLE
Adding support for datastore emulator in system test.

### PR DESCRIPTION
**Pull request by dhermes**

### CODE REVIEW

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -6,23 +6,23 @@ Contributing
 #. Make sure that your commit messages clearly describe the changes.
 #. Send a pull request.
 
-Here are some guidelines for hacking on gcloud-python.
+Here are some guidelines for hacking on ``gcloud-python``.
```
Comment (dhermes) [inline CONTRIBUTING.rst]: @tseaver I can undo this bit if you like, I just noticed that `gcloud-python` was plain text in a lot of places in this file so I did a find and replace.

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -39,17 +39,17 @@ using a Git checkout:
 Now your local repo is set up such that you will push changes to your GitHub
 repo, from which you can submit a pull request.
 
-- Create a virtualenv in which to install gcloud-python::
+- Create a virtualenv in which to install ``gcloud-python``::
 
    $ cd ~/hack-on-gcloud
-   $ virtualenv -ppython2.7 env
+   $ virtualenv --python python2.7 env
```
Comment (dhermes) [inline CONTRIBUTING.rst]: @tseaver Ditto here on unrelated changes. I was always like "WHAT is `-pp"? and have always been meaning to add the verbose flag.

File: system_tests/datastore.py [hunk 3]
```diff
@@ -271,7 +287,13 @@ def test_projection_query(self):
         self.assertEqual(catelyn_stark_dict,
                          {'name': 'Catelyn', 'family': 'Stark'})
 
-        catelyn_tully_entity = entities[3]
+        if EMULATOR_DATASET is None:
```
Comment (dhermes) [inline system_tests/datastore.py]: @pcostell This was surprising to me.

1. Why have we ever been able to rely on hardcoding the index returned from a projection query?
1. Why is the order different on the emulator than the backend?

File: system_tests/clear_datastore.py [hunk 4]
```diff
@@ -34,11 +33,15 @@
     'Post',
 ]
 TRANSACTION_MAX_GROUPS = 5
+if os.getenv('GCLOUD_NO_PRINT') == 'true':
+    PRINT_FUNC = lambda msg: None
+else:
+    PRINT_FUNC = print
```
Comment (tseaver) [inline system_tests/clear_datastore.py]: Can't work in Python2:  `print` is a keyword there.

Comment (dhermes) [inline system_tests/clear_datastore.py]: Not when `from __future__ import print_function` is used

File: system_tests/datastore.py [hunk 5]
```diff
@@ -35,8 +42,17 @@ class Config(object):
...
 
 def setUpModule():
-    client.DATASET = TESTS_DATASET
-    Config.CLIENT = datastore.Client()
+    if EMULATOR_DATASET is None:
+        client.DATASET = TESTS_DATASET
+        Config.CLIENT = datastore.Client()
+    else:
+        Config.CLIENT = datastore.Client(dataset_id=EMULATOR_DATASET)
+        populate_datastore.add_characters(client=Config.CLIENT)
```
Comment (tseaver) [inline system_tests/datastore.py]: Does this side need to initialize `client.DATASET = EMULATOR_DATASET`?

File: system_tests/clear_datastore.py (hunk 4 above)
Comment (tseaver) [inline system_tests/clear_datastore.py]: NM, I see you did the `from __future__ import print_function` above.

File: system_tests/datastore.py (hunk 5 above)
Comment (dhermes) [inline system_tests/datastore.py]: I elected to pass an explicit `dataset_id` to the constructor instead (to avoid side-effects)

**Comment by dhermes** [comment]
@tseaver FYI I added a 2nd commit here. I had some `pep8` errors but they weren't showing up because I forgot to use the status code in #1348. Fixing that (and the `pep8` violations) here, but if you think that fix should be standalone I'm happy to pull it into another PR.

**Comment by tseaver** [comment]
LGTM. Lint aside, I like the newer version better anyway.

BTW, `virtualenv -p python2.7` would have worked as well.

**Comment by dhermes** [comment]
Yes I realize, I just added the verbose one so there was no guessing for other confused folks.

Will wait on @pcostell before merging. (I think he'll probably be happy that our system tests run with the emulator.)

**Comment by dhermes** [comment]
Changed my mind. Merging so I don't forget. Will circle back to get @pcostell feedback later if we don't hear from him.

File: system_tests/datastore.py (hunk 3 above)
Comment (pcostell) [inline system_tests/datastore.py]: When answering this query, the ordering of `name` and `family` don't matter. My guess is the emulator is using the preferred index (lexographical) (family, name) where production is using an index that already exists (name, family).

This fits with the chosen indexes:

name      |  family     
--------- | ------------
Arya      |  Stark      
Bran      |  Stark      
Eddard    |  Stark      
Catelyn   |  Stark      
Catelyn   |  Tully      
Jon Snow  |  Stark      
Rickard   |  Stark    
Robb      |  Stark      
Sansa     |  Stark      

family  | name      
------- | --------
Stark   | Arya      
Stark   | Bran      
Stark   | Eddard    
Stark   | Catelyn   
Stark   | Jon Snow  
Stark   | Rickard   
Stark   | Robb      
Stark   | Sansa     
Tully   | Catelyn   

For the test I would definitely force an order:

    SELECT name, family from Characters WHERE __key__ HAS ANCESTOR(Key('Book', 'GoT')) ORDRE BY name, family;

Comment (dhermes) [inline system_tests/datastore.py]: 1. I will add the order to the query
1. Is there a way to add indexes to the emulator?

Thanks!

Comment (pcostell) [inline system_tests/datastore.py]: Yes, you can pass a `--data-dir` flag to tell it where to look. It will look for an index.yaml inside of the WEB-INF folder inside your `--data-dir`.

However, you should note that with `gcloud beta emulators` (vs gcd.sh) I don't think it is possible to tell the emulator to avoid building new indexes and fail instead.

Also huge :+1: to getting these working with the emulator!

Comment (dhermes) [inline system_tests/datastore.py]: w00t! Huge :+1: for making the emulator essentially trivial to work with.

Do you advise using our indices for the emulator? As can be seen in #1370, I am just populating the stark family tree at test case set-up and then deleting them on completion.

Comment (pcostell) [inline system_tests/datastore.py]: I think using your custom indexes is good because it makes your emulator usage match production usage.

But I think it's also not entirely necessary as long as you use queries which cannot choose which index to use.
